### PR TITLE
Enable powf*, exp*, log* intrinsics

### PR DESCRIPTION
**Pull request by tautschnig**
CBMC provides approximating implementations of these.

Resolves: #2966

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by jaisnan — COMMENTED** [review]
Would it be possible to add tests for exp, and sqrt as well similar to the powf test?

File: tests/kani/Intrinsics/Math/powf64.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    use super::*;
+
+    #[kani::proof]
+    fn verify_f() {
+        const LIMIT: u64 = 10;
+        let x: u64 = kani::any();
+        let y: u64 = kani::any();
+        // outside these limits our approximation may yield spurious results
+        kani::assume(x > LIMIT && x < LIMIT * 3);
+        kani::assume(y > LIMIT && y < LIMIT * 3);
```
Comment (adpaco-aws) [inline tests/kani/Intrinsics/Math/powf64.rs:20]: These assumptions are different to the ones in the test from #2966 - how did you find them?

File: docs/src/rust-feature-support/intrinsics.md [hunk 2]
```diff
@@ -185,10 +185,10 @@ nearbyintf64 | Yes | |
...
 nontemporal_store | No | |
 offset | Partial | Doesn't check [all UB conditions](https://doc.rust-lang.org/std/primitive.pointer.html#safety-2) |
-powf32 | No | |
-powf64 | No | |
-powif32 | No | |
-powif64 | No | |
+powf32 | Partial | Results are overapproximated |
+powf64 | Partial | Results are overapproximated |
+powif32 | Partial | Results are overapproximated |
+powif64 | Partial | Results are overapproximated |
```
Comment (adpaco-aws) [inline docs/src/rust-feature-support/intrinsics.md:191]: Only one of these is being tested.

**Review by adpaco-aws — CHANGES_REQUESTED** [review]
We need to have at least one test for each intrinsic that's being enabled here. I'd recommend separate PRs for each group of intrinsics.

File: tests/kani/Intrinsics/Math/powf64.rs (hunk 1 above)
Comment (tautschnig) [reply tests/kani/Intrinsics/Math/powf64.rs:20]: Trial an error. (But also: the assumptions in the original test seemed strange for they effectively looked for cases where numeric error was likely.)

**Comment by tautschnig** [comment]
> We need to have at least one test for each intrinsic that's being enabled here. I'd recommend separate PRs for each group of intrinsics.

Yes, more tests need to be added. But doing one PR per group is excessive overhead that I'm not willing to take on.

**Comment by rahulku** [comment]
Nudge. I believe there is a customer blocked on this.

File: docs/src/rust-feature-support/intrinsics.md (hunk 2 above)
Comment (tautschnig) [reply docs/src/rust-feature-support/intrinsics.md:191]: I have now added more tests.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 3]
```diff
@@ -456,8 +456,8 @@ impl<'tcx> GotocCtx<'tcx> {
             "log10f64" => unstable_codegen!(codegen_simple_intrinsic!(Log10)),
             "log2f32" => unstable_codegen!(codegen_simple_intrinsic!(Log2f)),
             "log2f64" => unstable_codegen!(codegen_simple_intrinsic!(Log2)),
-            "logf32" => unstable_codegen!(codegen_simple_intrinsic!(Logf)),
-            "logf64" => unstable_codegen!(codegen_simple_intrinsic!(Log)),
+            "logf32" => codegen_simple_intrinsic!(Logf),
+            "logf64" => codegen_simple_intrinsic!(Log),
```
Comment (feliperodri) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:460]: It's only missing tests for the log functions.

**Review by feliperodri — APPROVED** [review]
LGTM module log tests.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs (hunk 3 above)
Comment (tautschnig) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:460]: Tests added.

File: tests/kani/Intrinsics/Math/Arith/exp2.rs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
```
[the remark this answers is not in the document]
Comment (tautschnig) [reply tests/kani/Intrinsics/Math/Arith/exp2.rs:7]: Done.

File: tests/kani/Intrinsics/Math/Arith/exp.rs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#[kani::proof]
+fn verify_exp32() {
+    let two = 2.0_f32;
+    let two_sq = std::f32::consts::E * std::f32::consts::E;
+    let two_exp = two.exp();
+
+    assert!((two_sq - two_exp).abs() <= 1.0);
```
[the remark this answers is not in the document]
Comment (tautschnig) [reply tests/kani/Intrinsics/Math/Arith/exp.rs]: I've refined the bounds in all tests.

**Review by adpaco-aws — APPROVED** [review]

**Pull request merged**